StarDist model and demo

dkurt opened the pull request:
StarDist model from https://github.com/stardist/stardist

stardist_he | stardist_dsb2018
---|---
<img src="https://user-images.githubusercontent.com/25801568/111028579-8de2fc00-8408-11eb-8a7e-0f7979f39c63.png" width="256"> | <img src="https://user-images.githubusercontent.com/25801568/111028588-9affeb00-8408-11eb-9112-1f4d03fea212.png" width="256">
processing time: 12m38s | 6m02s

dkurt commented:
PR is ready, please review

dkurt commented:
@Wovchena, May I ask to review? Thanks in advance!

Wovchena reviewed (changes requested):
Please add models.lst file.

1. It is advised to avoid usage of these libraries reading images
    1.	They require additional installation step
    2. It will require to update CI to enable demo testing
    2.	Ideally, they should work for Mac as well, I don’t know if these libs are fine with that
    3. Is their license compatible with OMZ

    Is it reasonable to reimplement these libraries?
1. If this demo fits well into `object_detection_demo/python` idea, which seems to be, we require the demo to follow that approach. Some of the reasons:
    1. Current implementation of `stardist_demo` doesn’t catch exceptions in its threads, so it hangs when for example out of memory happens.
    2. `object_detection_demo/python` approach allows arbitrary combination of numbers of streams and numbers of infer requests.
    3. Having a dedicated `postprocess_thread_body` thread is questionable: I guess GIL doesn’t let you do much in parallel.
    4. `object_detection_demo/python` like demos separate model pre/postprocessing and the way they are executed.
1. It’s unpleasant to wait for processing of all tiles to get anything from the demo. The demo may crash for large images because of lack of memory. It seems tiles are processed independently. I suggest computing results on the fly, so the demo showed something before crashing :) The demo could show processed tiles while building `detections` dict or it could try preserving interactivity by processing tiles which are requested in first place.

CC @vladimir-dudnik, @IRDonch, @eizamaliev, @fzhar

dkurt commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,350 @@
+        if tile.shape[2] == 1:
+            tile = np.repeat(tile, 3, axis=-1)
+        else:
+            tile = cv.cvtColor(tile, cv.COLOR_RGB2BGR)
```
original tile is RGB and we converts it to BGR for visualization. That's why it's `COLOR_RGB2BGR`

dkurt commented:
> Having a dedicated postprocess_thread_body thread is questionable: I guess GIL doesn’t let you do much in parallel.

@Wovchena, `AsyncPipeline` does not utilize CPU in 100% for this pipeline (even with disabled NMS).

dkurt commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,350 @@
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument('-i', dest='input', required=True, help='Path to input image')
+    parser.add_argument('-m', dest='model', required=True,
```
Fixed

dkurt commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,350 @@
+    start_time = time.time()
+    iter_id = 0
+    while True:
+        tile, x, y = tiles_queue.get()
```
Done

dkurt commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,350 @@
+                # Try to keep balance between tiles queue and predictions queue. They should be stable and not empty.
+                # Too small tiles queue might give lower FPS because of not fully utilized device.
+                # Too big predictions queue may lead to huge memory usage and lower FPS because of it
+                # (increase confidence threshold to speed up postprocessing this way).
```
Removed due AsyncPipeline usage

dkurt commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,350 @@
+                # Too small tiles queue might give lower FPS because of not fully utilized device.
+                # Too big predictions queue may lead to huge memory usage and lower FPS because of it
+                # (increase confidence threshold to speed up postprocessing this way).
+                print('Processed {} tiles. Tiles in queue: {}. Predictions queue: {}. Average FPS: {:.2f}'.format(
```
Removed at all

dkurt commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,350 @@
+
+    # Render single tile at the time. User might navigate between tiles by arrows.
+    WIN_NAME = 'StarDist with OpenVINO'
+    cv.namedWindow(WIN_NAME, cv.WINDOW_NORMAL)
```
removed

dkurt commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,350 @@
+    KEY_LEFT = 81
+    KEY_UP = 82
+    KEY_RIGHT = 83
+    KEY_DOWN = 84
```
Removed. Now just rendering at the inference loop

dkurt commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,350 @@
+    tile_x = num_tiles_x // 2
+    tile_y = num_tiles_y // 2
+    while True:
+        preds = detections[(tile_y, tile_x)]
```
Fixed

dkurt commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,350 @@
+        cv.imshow(WIN_NAME, tile)
+
+        key = cv.waitKey()
+        if key == KEY_ESC:
```
Done, thanks!

dkurt commented on models/public/stardist_dsb2018/model.yml:
```diff
@@ -0,0 +1,31 @@
+  - $type: unpack_archive
+    format: zip
+    file: dsb2018_heavy_augment.zip
+model_optimizer_args:
```
Done

dkurt commented:
Removed threading completely (pre- and postprocessing optimizations let to run pipeline fast enough). `AsyncPipeline` API is used now. Moved tiles visualization to the inference thread.

Wovchena commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,373 @@
+logging.basicConfig(format='[ %(levelname)s ] %(message)s', level=logging.INFO, stream=sys.stdout)
+log = logging.getLogger()
+
+class OpenSlideImageReader:
```
```suggestion
class SlideImageReader:
```
and its usage

Wovchena commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,373 @@
+                level = i
+                break
+        if not level:
+            raise Exception('No level found for pixel size {}. Available levels are {}'.format(pixel_size, self.img.level_dimensions))
```
```suggestion
            raise RuntimeError('No level found for pixel size {}. Available levels are {}'.format(pixel_size, self.img.level_dimensions))
```

Wovchena commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,373 @@
+    return ids
+
+
+class StarDistModel:
```
Please remove `Model` suffix and move the class and corresponding functions like `normalize_percentile()` into common with other models.

Wovchena commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,373 @@
+                        help='Device to be utilized for deep learning inference')
+    parser.add_argument('-t', '--confidence_threshold', default=0.7, type=float,
+                        help='Probability threshold for detection')
+    parser.add_argument('--nms', dest='nms_threshold', default=0.5, type=float,
```
```suggestion
    parser.add_argument('-nms', '--nms_threshold', default=0.5, type=float,
```

Wovchena commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,373 @@
+        else:
+            return None
+
+    start = time.time()
```
Please reuse `PerformanceMetrics` https://github.com/openvinotoolkit/open_model_zoo/blob/7577edeaa83d8bc2d30f200cdc786fdfc1d4ddb1/demos/object_detection_demo/python/object_detection_demo.py#L248

It uses `perf_counter()` instead of `time()` to ensure time is steady, and along with reports of mean metrics after run, it reports a mean for last second on a screen.

Wovchena commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,373 @@
+                ESC_KEY = 27
+                # Quit.
+                if key in {ord('q'), ord('Q'), ESC_KEY}:
+                    break
```
Since the demo is just like others now, can you add `Presenter` https://github.com/openvinotoolkit/open_model_zoo/blob/7577edeaa83d8bc2d30f200cdc786fdfc1d4ddb1/demos/object_detection_demo/python/object_detection_demo.py#L293
drawing CPU and memory usage and handle key press for it here. And a demo arg to enable them at startup.

You can create an instance of `Presenter` out of main loop because the size of frame is known in advance.

Wovchena commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,373 @@
+                                          meta={'tile': tile, 'start_time': start_time})
+            tile_id = next_tile(tile_id)
+            if tile_id is None:
+                break
```
You shouldn't break here because it won't let to finish already submitted results.
We have the part starting with https://github.com/openvinotoolkit/open_model_zoo/blob/7577edeaa83d8bc2d30f200cdc786fdfc1d4ddb1/demos/object_detection_demo/python/object_detection_demo.py#L308
instead, but it looks it's simpler just to add `tile_id is not None` to `if detector_pipeline.is_ready():` in your case.

Wovchena commented on demos/stardist_demo/python/stardist_demo.py:
```diff
@@ -0,0 +1,373 @@
+        else:
+            # Wait for empty request
+            detector_pipeline.await_any()
+
```
https://github.com/openvinotoolkit/open_model_zoo/blob/7577edeaa83d8bc2d30f200cdc786fdfc1d4ddb1/demos/object_detection_demo/python/object_detection_demo.py#L306
is essential. Otherwise IE may not finalize gracefully.

Wovchena reviewed (changes requested):
Please add models.lst file.
Have you searched for other solutions regarding the question about libraries reading images?

dkurt commented:
See OpenVINO plugin for QuPath: https://github.com/dkurt/qupath-extension-openvino
